lib/kafka_producer.js

```diff
@@ -157,10 +244,54 @@ KafkaProducer.prototype.getWholeMsg = function getWholeMsg(topic, message, timeS
 
 KafkaProducer.prototype.close = function close(callback) {
     var self = this;
+    if (self.batching) {
+        self.flushEntireCache(callback);
+    }
     self.enable = false;
     if (self.restClient) {
         self.restClient.close();
     }
 };
 
+// The batch message payload should follow:
+// 4 BE bytes number of messages + 4 BE bytes size of message + actual message
+function MessageBatch(topic) {
+    var self = this;
+
+    self.payload = null;
+    self.timestamp = new Date().getTime();
+    self.numMessages = 0;
+    self.sizeBytes = 0;
+    self.topic = topic;
+}
+
+MessageBatch.prototype.addMessage = function addMessage(message) {
```

Comment:
what you want to do here is simply track the offset of what you've written so far.

i.e.

```js
var offset = self.currOffset;
var msgOffset = offset + 4;
// IMPORTANT: message must be string
var bytesWritten = self.cachedBuf.write(message, msgOffset) 
self.cachedBuf.writeUInt32BE(bytesWritten, offset);
self.offset += 4 + bytesWritten
```
```